rust/src/container.rs

```diff
@@ -0,0 +1,104 @@
+//! CLI exposing `ostree-rs-ext container`
+
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use std::convert::TryInto;
+
+use anyhow::{Context, Result};
+use structopt::StructOpt;
+
+#[derive(Debug, StructOpt)]
+enum Opts {
+    /// Import an ostree commit embedded in a remote container image
+    Import {
+        /// Path to the repository
+        #[structopt(long)]
+        repo: String,
+
+        /// Image reference, e.g. registry:quay.io/exampleos/exampleos:latest
+        imgref: String,
+    },
+
+    /// Print information about an exported ostree-container image.
+    Info {
+        /// Image reference, e.g. registry:quay.io/exampleos/exampleos:latest
+        imgref: String,
+    },
+
+    /// Export an ostree commit to an OCI layout
+    Export {
+        /// Path to the repository
+        #[structopt(long)]
+        repo: String,
+
+        /// The ostree ref or commit to export
+        rev: String,
+
+        /// Image reference, e.g. registry:quay.io/exampleos/exampleos:latest
+        imgref: String,
+    },
+}
+
+async fn container_import(repo: &str, imgref: &str) -> Result<()> {
+    let repo = &ostree::Repo::open_at(libc::AT_FDCWD, repo, gio::NONE_CANCELLABLE)?;
+    let imgref = imgref.try_into()?;
+    let (tx_progress, rx_progress) = tokio::sync::watch::channel(Default::default());
+    let target = indicatif::ProgressDrawTarget::stdout();
+    let style = indicatif::ProgressStyle::default_bar();
+    let pb = indicatif::ProgressBar::new_spinner();
+    pb.set_draw_target(target);
+    pb.set_style(style.template("{spinner} {prefix} {msg}"));
+    pb.enable_steady_tick(200);
+    pb.set_message("Downloading...");
+    let import = ostree_ext::container::import(repo, &imgref, Some(tx_progress));
+    tokio::pin!(import);
+    tokio::pin!(rx_progress);
+    loop {
+        tokio::select! {
+            _ = rx_progress.changed() => {
+                let n = rx_progress.borrow().processed_bytes;
+                pb.set_message(&format!("Processed: {}", indicatif::HumanBytes(n)));
+            }
+            import = &mut import => {
+                pb.finish();
+                println!("Imported: {}", import?.ostree_commit);
+                return Ok(())
+            }
+        }
+    }
+}
+
+async fn container_export(repo: &str, rev: &str, imgref: &str) -> Result<()> {
+    let repo = &ostree::Repo::open_at(libc::AT_FDCWD, repo, gio::NONE_CANCELLABLE)?;
+    let imgref = imgref.try_into()?;
+    let pushed = ostree_ext::container::export(repo, rev, &imgref).await?;
+    println!("{}", pushed);
+    Ok(())
+}
+
+async fn container_info(imgref: &str) -> Result<()> {
+    let imgref = imgref.try_into()?;
+    let info = ostree_ext::container::fetch_manifest_info(&imgref).await?;
+    println!("{} @{}", imgref, info.manifest_digest);
+    Ok(())
+}
+
+/// Main entrypoint for container
+pub fn entrypoint(args: &[&str]) -> Result<()> {
+    tokio::runtime::Builder::new_multi_thread()
```

Comment:
Going forward, it may be sensible to factor this further and add a https://docs.rs/tokio/1.6.0/tokio/runtime/struct.Runtime.html#method.shutdown_timeout on top.

Reply:
I think we discussed this in a zincati PR, my view is basically that the ideal case is we have no `SIGTERM` handler installed normally, and on system shutdown our process is unilaterally killed by the kernel without running through our code at all.  No need to call `free()` on a bunch of stuff, wait for threads or whatever.  It should all be "stateless".

xref https://github.com/nhorman/rng-tools/pull/72 and https://internals.rust-lang.org/t/should-rust-programs-unwind-on-sigint/13800/11

Reply:
That said rpm-ostreed does have a `SIGTERM` handler installed which was initially done to gracefully release our DBus name around automatic shutdown.

And we do exit our mainloop and return all the way out of main, so we'll end up invoking automatic cleanup functions.  But...only for the main thread!  Right now we aren't actually waiting for the transaction worker thread (if one exists) - you can easily see this by doing e.g. `rpm-ostree initramfs --enable && sleep 2 && systemctl stop rpm-ostreed` - which IMO is correct.  A whole part of the goal here is that upgrades should be killable at any time.  So waiting synchronously for them would unnecessarily delay things.